enhancements/network/static-ip-addresses-vsphere.md

```diff
@@ -0,0 +1,564 @@
+2. Add validation for the modified/added fields in the platform specification.
+3. For compute nodes, produce machine manifests with associated IP configuration.  
+
+Example of `machine` configured with IP configuration
+~~~yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: Machine
+metadata:
+  name: test-compute-1
+spec:
+  lifecycleHooks: {}
+  metadata: {}
+  providerSpec:
+    value:
+      numCoresPerSocket: 2
+      diskGiB: 60
+      snapshot: ''
+      userDataSecret:
+        name: worker-user-data
+      memoryMiB: 8192
+      credentialsSecret:
+        name: vsphere-cloud-credentials
+      network:
+        devices:
+          - networkName: lab
+            config:
+              interfaces:
+                - ipv4:
+                    address:
+                      - ip: 192.168.101.245
+                        prefix-length: 23
+              dns-resolver:
+                config:
+                  server:
+                    - 192.168.1.215
+              routes:
+                config:
+                  - next-hop-address: 192.168.100.1
+      metadata:
+        creationTimestamp: null
+      numCPUs: 2      
+      kind: VSphereMachineProviderSpec
+      workspace:
+        datacenter: testdc
+        datastore: datastore-1
+        folder: /testdc/vm/cluster-folder
+        resourcePool: /testdc/host/cluster1/Resources
+        server: test.vcenter.net
+      template: vm-template-rhcos
+      apiVersion: machine.openshift.io/v1beta1
+~~~
+4. For bootstrap and control plane nodes, modify vSphere terraform to convert nmstate to a VM guestinfo parameter
+for each VM to be created.
+
+As the assets are generated for the control plane and compute nodes, the slice of `host`s for each 
+node role will be used to populate `nmstate` information.  The number of `host`s must match the number of
+replicas defined in the associated machine pool.
+
+Additionally, each defined host may optionally define a failure domain.  This indicates that the associated `networkConfig` will be applied to a machine created in the indicated failure domain.
+
+
+##### Machine API
+- Modify vSphere machine controller to convert nmstate to VM guestinfo parameter
+- Introduce a new lifecycle hook called `preCreate`.
+- Modify [types_vsphereprovider.go](https://github.com/openshift/api/blob/master/machine/v1beta1/types_vsphereprovider.go) to support nmstate configuration. 
+
+
+###### IP Configuration of Machines
+The machine API `VSphereMachineProviderSpec.Network` will be modified to include a new type called `NetworkConfig`.  
+
+~~~go
+// NetworkDeviceSpec defines the network configuration for a virtual machine's
+// network device.
+type NetworkDeviceSpec struct {
+	// NetworkName is the name of the vSphere network to which the device
+	// will be connected.
+	NetworkName string `json:"networkName"`
+
+	// Config specifies an IP configuration to be applied upon cloning.
+	// +optional
+	Config *NetworkConfig `json:"config,omitempty"`
+}
+~~~
```

Comment:
Do we have a PR for these changes? If not, can you create one so that we can start the API review process there please

Reply:
We do, but it needs some attention https://github.com/openshift/api/pull/1334 .  I'll work on that today.

Reply:
A pull request has been started for the nmstate related API changes https://github.com/openshift/api/pull/1338 . I was torn on where the nmstate types should live so any feedback on that is much appreciated.

Reply:
This is a little different circumstance than when we implemented nmstate for baremetal, but in that case we intentionally made the nmstate data opaque to most of the components. The API for it is just a JSON blob of nmstate config that gets passed around until nmstate processes it. I'm not sure there's as much value to that here since an OpenShift component needs to process it at some point, but maybe it's a bit simpler if we don't try to represent the whole schema in the public API?

Reply:
perhaps you're right about that.  I'll investigate some options for this that don't involve adding it to the public API.